### DISCUSSION TITLE
MVP Endpoints

### DISCUSSION THREAD

**Discussion by TylerMatteo** [NYCPlanning/ae-zoning-api#21] [Ideas]
Let's sketch out which endpoints we need in addition to the `/boroughs` one already speced out in #9 !

* Individual tax lot endpoint - This could be `/tax-lots/<bbl>` or something more verbose that conveys the hierarchy of borough-block-lot, such as `/boroughs/<borough_id>/blocks/<block>/lots/<lot>`. @TangoYankee I think we settled on the simpler BBL approach last time we discussed this but open to discussion. We know that the frontend end will need to pan the map the a tax lot when the user looks up a borough, block, and lot. Depending on exactly how that interaction with Deck works, this endpoint will probably need to return geojson. In that case, we could append `/geojson` on the end, implying that an endpoint without it would include only the properties about the lot and not its geometry.
    * We also need to list out the zoning district classes for the zoning district that the lot resides in ([designs](https://www.figma.com/file/JRtcvgpkavWjBu7EF7vBYO/New-Tech-Stack-ZoLa-Mockups?type=design&node-id=342-27248&mode=design&t=Dpq3L1HwS0nupGi5-0). We could includes these in a nested property of the tax lot but I tend to prefer avoiding nested response structures in favor of making multiple API calls with flatter responses. If you take the multiple calls approach, we could have something like `/tax-lots/<bbl>/zoning-district/zoning-district-classes`
* List zoning district classes - Probably just `/zoning-district-classes`. We probably want this to return all of them in one go so that we can use that data to do things like render out the toggles for hiding/showing districts on the map. My initial thought was that this would be an object "indexed" by the class ideas but on second thought I think it probably makes more sense to just return an array of zoning district class objects and convert it to the nested object structure on the front end if needed.
* List of land uses - `/land-uses` - Same deal as zoning district classes, but for land uses
* Individual zoning district class endpoint - `/zoning-district-classes/<id>` This endpoint supports the functionality of what happens when a user clicks on an individual zoning district on the map. We'll have the zoning district's class id via Deck when the user clicks on a polygon but stuff like the description is associated with that class. If we decide to just get a reference to the class from the list of classes on the front end, we may not need this

## Vector tile endpoints
* `/tax-lots/<z>/<x>/<y>.pbf`
* `/zoning-districts/<z>/<x>/<y>.pbf`
These endpoints will redirect traffic to S3 storage but I'm thinking maybe we don't include them in the `openapi.yaml` for now? There are ways to represent protocol buffer-encoded responses in OAS but it might something we should punt on for now.

**Comment by TangoYankee** [comment]
## General rules
- Responses with more than one object are returned as an array of objects. ie) zoning classes will be an array and not an indexed object
- Prefer shallow routes that can be composed with multiple calls to get the required data.
  - For example, the land use for a tax-lot would be acquired with a call to `tax-lot/<bbl>` to get the foreign key to land use. This key would then be used in a call to `land-use/<id>`
  - Many to many relationships are the exception, as the foreign keys would not actually be returned in the initial call. Instead, the relationship is represented with a nested path. For example, the classes for a zoning-district would be at `zoning-districts/<id>/classes'

## Breaking the general rules
- returning nested objects for many to one relationships when it provides a simpler request for a known front end need
   - For example, change `tax-lots/<uuid>` to return its land use and borough as nested objects 

## Summary of endpoints
 - [ ] `land-uses`
   - all land use fields of all land uses
 - [ ] ~~`land-uses/<id>`~~
    - all land use fields of the specified land use
 - [x] `boroughs` *(already written)*
   - all fields of all boroughs
 - [ ] ~~`boroughs/<id>`~~
    - all fields of the specified borough
- [ ] `tax-lots/<bbl>`
   - non-spatial tax-lot fields
     - land use value should be a nested land use object
     - borough value should be a nested borough object
- [ ] `tax-lots/<bbl>/zoning-districts`
   - non-spatial fields of zoning districts that spatially intersect the tax lot
- [ ] `tax-lots/<bbl>/zoning-districts/classes`
   - all fields of zoning district classes for zoning districts that spatially intersect the tax lot 
- [ ] ~~`tax-lots/<bbl>/zoning-districts/geojson`~~ *(Not necessary for the pilot functionality but would be a great demonstration for MikroOrm with PostGIS)*
   - non-spatial fields of zoning districts that spatially intersect the tax lot as properties
   - spatial intersection of zoning districts with tax lots in wgs84 as geometry  
- [ ] `tax-lots/<bbl>/geojson`
   - non-spatial fields as properties
     - land use value should be a nested land use object
     - borough value should be a nested borough object
   - wgs84 as geometry
 - [ ] `zoning-district-classes`
   - all fields of all zoning-district-classes
 - [ ] `zoning-district-classes/category-colors`
   - list of color and the class category it represents
 - [ ] `zoning-district-classes/<id>`
    - all fields of the specified zoning-district-class 
 - ~~`zoning-districts`~~
   - non-spatial fields of all zoning districts
 - [ ] `zoning-districts/<uuid>`
   - non-spatial fields of specified zoning district
 - [ ] `zoning-districts/<uuid>/classes`
   - all fields of all zoning-district-classes associated with the specified zoning district 
   - ~~query parameter to filter based on category~~
  - ~~`zoning-districts/geojson`~~
   - non-spatial fields as property
   - wgs84 as geometry
 - [ ] ~~`zoning-districts/<uuid>/geojson`~~
   - non-spatial fields as property
   - wgs84 as geometry

**Reply by TangoYankee** [reply]
@TylerMatteo I want to highlight the `zoning-district-classes/category-colors` as a way for the db to manage which color represents its category for all classes within that category.

**Reply by TylerMatteo** [reply]
I was thinking we would just grab the color out of the list of zd classes on the frontend but if you think it's worth having another endpoint, fine by me.

What's are the use cases for `zoning-districts` and `zoning-districts/geojson`?

**Reply by TangoYankee** [reply]
> I was thinking we would just grab the color out of the list of zd classes on the frontend but if you think it's worth having another endpoint, fine by me.

Yeah! Let's experiment and review the experience afterwards

> What's are the use cases for `zoning-districts` and `zoning-districts/geojson`

The abstract answer is: *to give users access to the data for a use we may not imagine but they may still desire*. You may retort *well, why provide such a route for zoning districts but not tax lots?*. I do think we should also provide an equivalent tax-lot route. But, there are so many tax lots that we would need pagination to avoid freezing the applications. Zoning districts should at least be small enough that an unpaginated request would only be slow, instead of breaking.

The practical is answer is: *we don't need them for the pilot and so we can exclude them*

I crossed them off the list

**Reply by TylerMatteo** [reply]
Gotcha, yeah let's keep this list specifically the ones we need for the demo. I was going to bring up the can of worms that is pagination but luckily we don't have to cross that particular bridge yet.

**Reply by TangoYankee** [reply]
>  let's keep this list specifically the ones we need for the demo

Hmm, should we also shelve `tax-lots/<bbl>/zoning-districts/geojson`?

**Reply by TylerMatteo** [reply]
@TangoYankee Other thoughts:
* There may be a couple cases where we can break out "don't nest" data rule for the sake of pragmatism, particularly in cases where the thing we're nesting is just an object with a few properties, not a list of objects. Examples of endpoints we wouldn't need if we did this include:
    * `/boroughs/<id>` - I think we only need the borough details for displaying tax lot details so we could put a `Borough` record under `borough` on the tax lot properties that come back from `/tax-lots/<bbl>`
    * `/land-uses/<id>` - Same deal as boroughs for tax lots
    * ^ This being said, this is a slippery slope so I'm cool with taking the puritan approach for now, this approach just saves us having to make sequential, potentially clunky, network requests on the frontend
* Scratch other endpoints we don't need for the demo, I think that just applies to `tax-lots/<bbl>/zoning-districts/geojson` now.

**Comment by TangoYankee** [comment]
- [ ] `land-uses`
   - all land use fields of all land uses
 - [x] `boroughs` *(already written)*
   - all fields of all boroughs
- [ ] `tax-lots/<bbl>`
   - non-spatial tax-lot fields
     - land use value should be a nested land use object
     - borough value should be a nested borough object
- [ ] `tax-lots/<bbl>/zoning-districts`
   - non-spatial fields of zoning districts that spatially intersect the tax lot
- [ ] `tax-lots/<bbl>/zoning-districts/classes`
   - all fields of zoning district classes for zoning districts that spatially intersect the tax lot 
- [ ] `tax-lots/<bbl>/geojson`
   - non-spatial fields as properties
     - land use value should be a nested land use object
     - borough value should be a nested borough object
   - wgs84 as geometry
 - [ ] `zoning-district-classes`
   - all fields of all zoning-district-classes
 - [ ] `zoning-district-classes/category-colors`
   - list of color and the class category it represents
 - [ ] `zoning-district-classes/<id>`
    - all fields of the specified zoning-district-class 
 - [ ] `zoning-districts/<uuid>`
   - non-spatial fields of specified zoning district
 - [ ] `zoning-districts/<uuid>/classes`
   - all fields of all zoning-district-classes associated with the specified zoning district

**Reply by TangoYankee** [reply]
@TylerMatteo I made a consolidated list. Any more changes before we start making tickets? Also, the following comment is how I see the work being divided. I figured that most of the work is in making schemas to reference. So, the routes should be split based on what schema they return.

**Reply by TangoYankee** [reply]
## Ticket A
Route that returns zoning district (blocks tickets B and C because it creates the zd schema)
 - [ ] `zoning-districts/<uuid>`
   - non-spatial fields of specified zoning district

## Ticket B (Blocked by ticket A)
These routes return land use, borough, zoning district and/or tax lot schemas
- [ ] `land-uses`
- [ ] `tax-lots/<bbl>`
   - non-spatial tax-lot fields
     - land use value should be a nested land use object
     - borough value should be a nested borough object
- [ ] `tax-lots/<bbl>/geojson`
   - non-spatial fields as properties
     - land use value should be a nested land use object
     - borough value should be a nested borough object
   - wgs84 as geometry
- [ ] `tax-lots/<bbl>/zoning-districts`
   - non-spatial fields of zoning districts that spatially intersect the tax lot
  
## Ticket C (Blocked by ticket A)
These routes return zoning district or zoning district class schemas
- [ ] `tax-lots/<bbl>/zoning-districts/classes`
   - all fields of zoning district classes for zoning districts that spatially intersect the tax lot 
 - [ ] `zoning-district-classes`
   - all fields of all zoning-district-classes
 - [ ] `zoning-district-classes/category-colors`
   - list of color and the class category it represents
 - [ ] `zoning-district-classes/<id>`
    - all fields of the specified zoning-district-class 
 - [ ] `zoning-districts/<uuid>/classes`
   - all fields of all zoning-district-classes associated with the specified zoning district

**Reply by TangoYankee** [reply]
The implementation tickets will follow the same breakdown as these documentation tickets, with each implementation ticket blocked by its corresponding documentation ticket
